File: src/medium/#852. Peak Index in a Mountain Array.js
// https://leetcode.com/problems/peak-index-in-a-mountain-array/description/

function peakIndexInMountainArray(arr) {
  return binSearch(arr, 0, arr.length);
}

function binSearch(arr, start, end) {
  let middleIndex = Math.floor((start + end) / 2);

  if (arr[middleIndex - 1] < arr[middleIndex] && arr[middleIndex] > arr[middleIndex + 1]) {
    return middleIndex;
  }

  if (arr[middleIndex] < arr[middleIndex + 1]) {
    return binSearch(arr, middleIndex + 1, end);
  } else {
    return binSearch(arr, start, middleIndex - 1);
  }
}

console.log(peakIndexInMountainArray([0, 2, 1, 0]));

File: src/medium/#852. Peak Index in a Mountain Array.py
import math


def peakIndexInMountainArray(arr):
    return bin_search(arr, 0, len(arr)-1)


def bin_search(arr, start, end):
    middleIndex = math.floor((start + end) / 2)

    if start > end:
        return -1

    if arr[middleIndex] > arr[middleIndex + 1] and arr[middleIndex] > arr[middleIndex - 1]:
        return middleIndex

    # не обязательно проверять, что текущий элемент меньше предыдущего
    if arr[middleIndex] > arr[middleIndex+1]:
        return bin_search(arr, start, middleIndex-1)
    else:
        return bin_search(arr, middleIndex+1, end)
